### PR TITLE
[ELF] --no-rosegment: don't mark read-only PT_LOAD segments executable

### PR DESCRIPTION
**Pull request by MaskRay**
Once we move `.lrodata` after .bss (#78521), or if we use `SECTIONS`
commands, certain read-only sections may be in their own PT_LOAD, not in
the traditional "text segment". Current --no-rosegment code may
unnecessarily mark read-only PT_LOAD executable. Fix it.

### CODE REVIEW

File: lld/ELF/Writer.cpp [hunk 1]
```diff
@@ -2465,13 +2460,22 @@ SmallVector<PhdrEntry *, 0> Writer<ELFT>::createPhdrs(Partition &part) {
     // so when hasSectionsCommand, since we cannot introduce the extra alignment
     // needed to create a new LOAD)
     uint64_t newFlags = computeFlags(sec->getPhdrFlags());
+    // When --no-rosegment is specified, RO and RX sections are compatible.
+    uint32_t diff = flags ^ newFlags;
+    if (config->singleRoRx && !(newFlags & PF_W))
+      diff &= ~PF_X;
```
Comment (smithp35) [inline lld/ELF/Writer.cpp:2466]: There's an interesting conflict here between singleRoRx and execute-only, could be worth making these options incompatible in Driver.cpp as singleRoRx breaks execute-only.

This isn't a regression caused by this patch though so could be done in a different patch.

**Review by smithp35 — COMMENTED** [review]
I've got some small comments, but no objections should anyone else want to approve.

File: lld/ELF/Writer.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/Writer.cpp:2466]: The initial `--execute-only` patch (https://reviews.llvm.org/D49456) has an error `--execute-only does not support intermingling data and code`. 

It also has this error `"--execute-only and -no-rosegment cannot be used together"`. (The `!Script->HasSectionsCommand` condition it uses should be wrong as hasSectionsCommand is not set in Driver.cpp yet)

File: lld/ELF/Writer.cpp [hunk 2]
```diff
@@ -2451,7 +2446,7 @@ SmallVector<PhdrEntry *, 0> Writer<ELFT>::createPhdrs(Partition &part) {
     // Segments are contiguous memory regions that has the same attributes
     // (e.g. executable or writable). There is one phdr for each segment.
     // Therefore, we need to create a new phdr when the next section has
-    // different flags or is loaded at a discontiguous address or memory region
+    // compatible flags or is loaded at a discontiguous address or memory region
```
[the remark this answers is not in the document]
Comment (MaskRay) [reply lld/ELF/Writer.cpp:2449]: Thanks for the catch. Fixed the typo

**Review by smithp35 — APPROVED** [review]
LGTM on my side, thanks for the updates.

**Review by aeubanks — APPROVED** [review]

**Pull request merged**